src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/PkcsPalOpenSsl.Encrypt.cs

```diff
@@ -69,15 +70,141 @@ internal sealed partial class PkcsPalOpenSsl : PkcsPal
+            // EnvelopedData ::= SEQUENCE {
+            //     version CMSVersion,
+            //     originatorInfo[0] IMPLICIT OriginatorInfo OPTIONAL,
+            //     recipientInfos RecipientInfos,
+            //     encryptedContentInfo EncryptedContentInfo,
+            //     unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL }
+            //
+            // As well as it remarks that "If unprotectedAttrs is present, then version shall be 2."
+
+            byte[][] contentType = originalMessage.ReadAndSplitNextEncodedValue();
+            DerSequenceReader envelopedDataReader = originalMessage.ReadExplicitContextSequence();
+
+            int versionNumber = envelopedDataReader.ReadInteger();
+
+            if (versionNumber > 4)
+                throw new PlatformNotSupportedException(SR.Cryptography_Cms_VersionNumberPlatformNotSupported);
+
+            if (versionNumber == 0)
+                versionNumber = 2;
+
+            byte[] version = ConcatenateArrays(DerEncoder.SegmentedEncodeUnsignedInteger(new byte[] { (byte)versionNumber }));
+
+            // We don't need to parse the OriginatorInfo, RecipientInfos, or EncryptedContentInfo, so
+            // we can just extract it to concatenate it.
+            byte[] remainingDataOfSequence = envelopedDataReader.ReadRemainingData();
+
+            byte[][] envelopedData = DerEncoder.ConstructSegmentedExplicitSequenceFromPayload(
+                0 /* Context number */,
+                version,
+                remainingDataOfSequence,
+                EncodeUnprotectedAttributes(unprotectedAttributes));
+
+            return DerEncoder.ConstructSequence(
+                contentType,
+                envelopedData);
+        }
+
+        private static byte[] ConcatenateArrays(params byte[][] data)
+        {
+            int length = 0;
+            foreach (byte[] innerData in data)
+            {
+                length += innerData.Length;
+            }
+
+            byte[] concatenatedArray = new byte[length];
+
+            int offset = 0;
+            foreach(byte[] innerData in data)
+            {
+                Buffer.BlockCopy(innerData, 0, concatenatedArray, offset, innerData.Length);
+                offset += innerData.Length;
+            }
+
+            return concatenatedArray;
+        }
+
+        private byte[] EncodeUnprotectedAttributes(CryptographicAttributeObjectCollection unprotectedAttributes)
+        {
+            // In EnvelopedData UnprotectedAttributes are included as:
+            //
+            // unprotectedAttrs[1] IMPLICIT UnprotectedAttributes OPTIONAL
+            //
+            // UnprotectedAttributes ::= SET SIZE (1..MAX) OF Attribute
+            
+            byte[][][] setOfAttrs = new byte[unprotectedAttributes.Count][][];
+
+            for (int i = 0; i < unprotectedAttributes.Count; i++)
+            {
+                setOfAttrs[i] = EncodeAttribute(unprotectedAttributes[i]);
```

Comment:
the setOfAttrs payload values are also a SET OF, so they need to be sorted, too.